### PR TITLE
[ci] upload artifacts of production, rolling or daily

### PR DESCRIPTION
**Pull request by saw-jan**
## Description
This is a POC for deploying docker and binary build artifacts according to https://github.com/owncloud/ocis/issues/9264.

**What is known:**
- Production release docker tags: `owncloud/ocis:x.x.x`
- Rolling release docker tags: `owncloud/ocis-rolling:x.x.x` _(including production tags)_
- Production binary builds go to: `stable` subfolder
- Rolling binary builds go to: `rolling` subfolder
- Daily binary builds go to: `daily` subfolder

**Q?:**
- What kind of builds go to `testing` subfolder? Pre-releases - alpha, beta, rc?
- Docker tag for daily builds? `owncloud/ocis-rolling:latest`?
- What should happen to previous docker tags and binary builds? (move to respective docker repo and binary subfolders?) - not related to this PR though

**PR implementation:** tag release type from https://github.com/owncloud/ocis/issues/9264#issue-2318967021
- `commit push`:
  - docker-arm64-daily
  - docker-amd64-daily
  - manifest-daily
  - readme-daily
  - binaries-linux-daily
  - binaries-darwin-daily

- `v6.0.0-rc.1`:
  - docker-arm64-rolling
  - docker-amd64-rolling
  - manifest-rolling
  - readme-rolling
  - binaries-linux-testing
  - binaries-darwin-testing

- `v6.0.0`:
  - docker-arm64-rolling
  - docker-amd64-rolling
  - manifest-rolling
  - readme-rolling
  - binaries-linux-rolling
  - binaries-darwin-rolling

- `v7.0.0-rc.1`:
  - docker-arm64-rolling
  - docker-amd64-rolling
  - manifest-rolling
  - readme-rolling
  - docker-arm64-production
  - docker-amd64-production
  - manifest-production
  - readme-production
  - binaries-linux-testing
  - binaries-darwin-testing

- `v7.0.0`:
  - docker-arm64-rolling
  - docker-amd64-rolling
  - manifest-rolling
  - readme-rolling
  - docker-arm64-production
  - docker-amd64-production
  - manifest-production
  - readme-production
  - binaries-linux-rolling
  - binaries-darwin-rolling
  - binaries-linux-production
  - binaries-darwin-production

## Related Issue
- https://github.com/owncloud/ocis/issues/9264

## How Has This Been Tested?
- :exclamation: END RESULT NOT TESTED

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by micbar** [comment]
Good approach!

Please keep in mind, that all `production` tags are also `rolling` tags.

**Comment by saw-jan** [comment]
>Please keep in mind, that all `production` tags are also `rolling` tags.

@micbar I have some questions regarding it in the PR description.
https://github.com/owncloud/ocis/pull/9364#issue-2348678169

Also, I have added about which pipelines will run on which type of events.

File: .drone.star [hunk 1]
```diff
@@ -1259,20 +1263,36 @@ def logTracingResults():
...
+    docker_repos = []
+    build_type = "daily"
+
+    # dockerhub repo
+    #  - "owncloud/ocis-rolling"
+    repo = ctx.repo.slug + "-rolling"
+    docker_repos.append(repo)
+
+    # production release repo
+    if ctx.build.event == "tag" and ctx.build.ref.replace("refs/tags/v", "") in PRODUCTION_RELEASE_TAGS:
```
Comment (micbar) [inline .drone.star:1275]: Can we use a startswith logic to be able to have production release tags like 5.0 which match with all patch releases?

Comment (saw-jan) [reply .drone.star]: yeah, totally forgot about that one

Comment (saw-jan) [reply .drone.star]: hmm, another question.
If I set `5.0` as a production tag, then `5.0.9-rc.1` will also be counted as the production release. Is it fine? If not then we need to set some logic to say that all `5.0` but not `-rc`, `-beta`, `-alpha` unless specified as production release

Comment (micbar) [reply .drone.star]: That is fine. Prereleases of the respective channels should also be published.

File: .drone.star [hunk 2]
```diff
@@ -1,6 +1,12 @@
 """oCIS CI definition
 """
 
+# Production release tags
+# NOTE: need to be updated if new production releases are determined
+# - follow semver
+# - omit 'v' prefix
+PRODUCTION_RELEASE_TAGS = ["5.0", "7.0.0"]
```
Comment (saw-jan) [inline .drone.star:8]: This treats the following tags as production:
- `5.0.x`, `5.0.x-alpha.x`, `5.0.x-beta.x`, `5.0.x-rc.x` (`5.0<anything>`)
- `7.0.0`, `7.0.0-alpha.x`, `7.0.0-beta.x`, `7.0.0-rc.x` (`7.0.0<anything>`)

Comment (micbar) [reply .drone.star:8]: Perfect.

**Comment by saw-jan** [comment]
The PR is complete from my side and the things remaining is to test if it does the required job. I don't know how to confirm it. But I tested pipelines generation using:
```
drone starlark --max-execution-steps 100000000 --build.event="tag" --build.ref="refs/tags/v6.0.0" --repo.slug="owncloud/ocis" --repo.name="ocis"

drone starlark --max-execution-steps 100000000 --build.event="tag" --build.ref="refs/tags/v5.0.5" --repo.slug="owncloud/ocis" --repo.name="ocis"

drone starlark --max-execution-steps 100000000 --build.event="pull_request" --build.ref="refs/pull/123" --repo.slug="owncloud/ocis" --repo.name="ocis"
```

File: ocis/docker/manifest.tmpl [hunk 3]
```diff
@@ -1,16 +1,16 @@
-image: owncloud/ocis:{{#if build.tag}}{{trimPrefix "v" build.tag}}{{else}}latest{{/if}}
+image: owncloud/ocis-rolling:{{#if build.tag}}{{trimPrefix "v" build.tag}}{{else}}latest{{/if}}
```
Comment (saw-jan) [inline ocis/docker/manifest.tmpl:1]: I have no idea how docker manifest works so I wonder if this is correct

Comment (micbar) [reply ocis/docker/manifest.tmpl:1]: this is fine

**Comment by micbar** [comment]
Rebased it.

**Review by micbar — APPROVED** [review]

**Pull request merged**